src/utils/colorManipulation.test.ts

```diff
@@ -0,0 +1,266 @@
+        },
+        {
+          row: 0,
+          col: 50,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+        {
+          row: 25,
+          col: 0,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+        {
+          row: 25,
+          col: 50,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+        {
+          row: 50,
+          col: 0,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+        {
+          row: 50,
+          col: 50,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+        {
+          row: 75,
+          col: 0,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+        {
+          row: 75,
+          col: 50,
+          width: 50,
+          height: 25,
+          color: new Uint8ClampedArray([255, 0, 0, 255]),
+        },
+      ];
+
+      const areDivisionsEqual = (d1: Division, d2: Division) =>
+        d1.width === d2.width &&
+        d1.height === d2.height &&
+        d1.row === d2.row &&
+        d1.col === d2.col;
+
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      const actual = computeDivisions(
+        frame,
+        videoDimensions,
+        horizontalDivisions,
+        verticalDivisions,
+      );
+      expected.forEach((expectedDiv) => {
+        expect(
+          actual.some((actualDiv) => areDivisionsEqual(actualDiv, expectedDiv)),
+        ).toBe(true);
+      });
+    });
+  });
+
+  describe("findVideoPositionOnCanvas", () => {
+    it("find the correct video boundaries for a centered video", () => {
+      const videoDimensions: Dimensions = { height: 50, width: 50 };
+
+      const expected: Position = {
+        top: 25,
+        right: 75,
+        bottom: 75,
+        left: 25,
+      };
+
+      const actual = findVideoPositionOnCanvas(
+        { height: canvas.height, width: canvas.width },
+        videoDimensions,
+      );
+      expect(actual).toEqual(expected);
+    });
+  });
+
+  describe("regionConvolution", () => {
+     it("convolves a region based on the kernel size given", () => {
+      canvas.width = 12;
+      canvas.height = 12;
+
+      //cut the left half to be red
+      ctx.fillStyle = "rgb(255,0,0)";
+      ctx.fillRect(0, 0, canvas.width / 2, canvas.height);
+
+      //cut the right half to be blue
+      ctx.fillStyle = "rgb(0,0,255)";
+      ctx.fillRect(canvas.width / 2, 0, canvas.width, canvas.height);
+
+      const frame = ctx.getImageData(0,0, canvas.width, canvas.height);
+      expect(Array.from(frame.data)).toStrictEqual(Array.from(EXPECTED_TEST_CASE_1_BEFORE_CONVOLVE));
```

Comment:
can you leave a brief comment explaining why we need this `Array.from` approach

Reply:
perhaps you are messing this with the polyfills that we did for the canvas package? for the description i stated in the comments, the ImageData object (frame.data) was being read in this dictionary-like syntax when being run in the test, thats why we were getting the issue. we had to convert them both into normal arrays and check if they values are the same there.

Reply:
hmm i think you've got the right answer, but for the wrong reason. you're correct in saying that `frame.data` came out in a weird object-like way, but it wasn't jest's fault. like you called out when we paired, the canvas package implementation of `ImageData` doesn't match that of the DOM anymore. we know already that the `colorSpace` readonly prop is missing from the `canvas` package when it's in the DOM's `ImageData`.  even though we're able to create our own polyfill for the DOM's `ImageData` with the `BrowserCompatibleImageData`, it's likely still that the internal structure of the canvas `ImageData` doesn't match that of the DOM's and **thats** whats causing the mismatch

lmk if that's clear or if i should clarify certain parts

Reply:
i think i understand it; the canvas package handles things differently, and the frame.data (which is a uint8clampedarray) is being handled differently, so just comparing if the ARRAYS are equal to each other is not going to work (arr1 === arr2 returns false; arr[1] === arr[2] returns true). that is why we had to transform them into normal javascript arrays by doing Array.from